Jetpack Partner portal (Billing): support no invoice available response

kallehauge opened the pull request:
## Changes proposed in this Pull Request

This PR aims to add support for the `no_billing_invoice_available` error code in the billing dashboard added by D62037-code.

We wish to support this error in the front-end because everything has actually worked up until this point and is not a direct error (they're not missing any settings or any connections went wrong) but we - Automattic - have not started creating their invoice yet.

The current implementation just displays loading indicators with a red notification which communicates that something went wrong and could potentially lead the guest to be confused about what to do to fix this error.

## Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

### Easy way

* Add `return new WP_Error( 'no_billing_invoice_available', 'Unable to retrieve upcoming invoice', '' );` to the beginning of `Jetpack_Licensing\Repository\License::get_billing_details()` to imitate the response without creating a new partner without an upcoming invoice.
* Start the `cloud.jetpack.com` site locally (`npm start start-jetpack-cloud`)
* Go to the Partner Portal (http://jetpack.cloud.localhost:3000/partner-portal)
* Select any of your keys (Testing or Production)
* Go to the Billing page and wait for the page to load

You should now see a grey notification saying that _we have not started calculating the upcoming invoice yet and the statistics might be wrong._ and the billing page itself should show zeros as default value.

### 1:1 Test

This requires some Jetpack partnership knowledge to be able to test since you have to create a new partner.

* Create a new partner (the partner has to be new since they should not have an upcoming invoice)
* Start the `cloud.jetpack.com` site locally (`npm start start-jetpack-cloud`)
* Go to the Partner Portal (http://jetpack.cloud.localhost:3000/partner-portal)
* Select any of your keys (Testing or Production)
* Go to the Billing page and wait for the page to load

You should now see a grey notification saying that _we have not started calculating the upcoming invoice yet and the statistics might be wrong._ and the billing page itself should show zeros as default value.
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

## Screenshots

**Before:**
The old way we handled the "no upcoming invoice" response.

![Screenshot 2021-06-07 at 14 03 50](https://user-images.githubusercontent.com/3846700/121013536-3a7b1280-c799-11eb-8e4b-3aa292d501d0.png)

**After:**
The suggested way we should handle the "no upcoming invoice" response.

![Screenshot 2021-06-07 at 19 39 45](https://user-images.githubusercontent.com/3846700/121066034-c016b680-c7c9-11eb-93ce-dcd55e94dfa6.png)

---


**Loading:**
Some places had double `<TextPlaceholder>` indicators which has been simplified during our refactor to display a `-` default value when the page is done loading.
We do this because it would have made the code unnecessarily more difficult to read/maintain and it didn't make a uniform loading experience to begin with. 

![Screenshot 2021-06-07 at 19 41 02](https://user-images.githubusercontent.com/3846700/121066091-cefd6900-c7c9-11eb-84ad-5086ee5ab70a.png)

## Related

1187494150150258-as-1200408426367592
1187494150150258-as-1200359063069386
D62037-code

kallehauge committed:
Partner portal: support no invoice available respose in billing dashboard

kallehauge committed:
Partner Portal: add test for no invoice available billing respose

atanas-dev commented on client/state/partner-portal/licenses/hooks/use-billing-dashboard-query.ts:
```diff
@@ -114,4 +132,35 @@ export default function useBillingDashboardQuery< TError = unknown >(
+				unassigned: 0,
+			},
+		};
+	}
```
We shouldn't need to mess with react-query internals such as isSuccess, isError etc.. We can either:
- Act on the error in the components that use the query (e.g. render dashes that signify the lack of data)
- ... or catch the rejected promise in queryBillingDashboard and stub in empty data there given the right error code or just re-reject (react-query will take it from there)

kallehauge replied:
It's me being stuck in thinking it's not actually an error but a `204` response and using `403` as a way to have better error logging 😅 

I have done the following:

* Generalised the way we handle `<TextPlaceholder>` so we only display it during `isLoading` state.
* Provided `-` as a default value for "Not successful" responses.**
* Removed the empty response in the `react-query` and use the updated default error handling.

_** We use the `-` for all "Not successful" responses to indicate that we are done loading the page and nothing more is going to happen._

atanas-dev commented on client/state/partner-portal/licenses/hooks/use-billing-dashboard-query.ts:
```diff
@@ -99,12 +99,30 @@ export default function useBillingDashboardQuery< TError = unknown >(
+					dispatch(
+						plainNotice(
+							translate(
+								'We have not started calculating your upcoming invoice yet. The statistics might not be correct.'
```
I think we can improve the wording. How about this as an example:
```suggestion
								'Your upcoming invoice is being prepared and will be available soon.'
```

kallehauge replied:
I like the updated wording since we use `-` instead of `0` as the fallback/default value we no longer need the `The statistics might not be correct.` disclaimer anyway. Thank you for this!

atanas-dev commented on client/state/partner-portal/licenses/test/hooks.js:
```diff
@@ -320,6 +320,43 @@ describe( 'useBillingDashboardQuery', () => {
 		expect( result.current.data ).toEqual( formattedStub );
 	} );
 
+	it( 'dispatches notification on no invoice available', async () => {
```
If we're going to stub in empty data we should have a test case or assertion that it gets returned.

kallehauge replied:
We no longer stub empty data

kallehauge committed:
JPPHP: display a value instead of placeholders during billing errors

We do this to indicate the we are done loading the request and the data in the placeholders are not going to be filled with relevant data.

kallehauge committed:
JPPHP: use the default error presentation during missing upcoming invoice

kallehauge commented:
**_Edit:_ Unrelated test no longer fails**

---

There is 1 unrelated test in [E2E tests (desktop) (Web app)](https://teamcity.a8c.com/viewLog.html?buildId=6271940&buildTypeId=calypso_RunCalypsoE2eDesktopTests) ([WPCOM] Invites - New user as Editor) which has been failing randomly through today due to timeouts and all related tests has been successfully completed.

kallehauge committed:
JPPHP: remove unnecessary variable in useBillingDashboardQuery

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/billing-details/index.tsx:
```diff
@@ -80,31 +80,19 @@ export default function BillingDetails(): ReactElement {
-							<span className="billing-details__line-item-meta">
-								<TextPlaceholder />
-							</span>
+							{ billing.isLoading ? <TextPlaceholder /> : '-' }
```
We use `<Gridicon icon="minus" />` in other parts of the Partner Portal rather than a literal `-`.

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/billing-details/index.tsx:
```diff
@@ -119,12 +107,12 @@ export default function BillingDetails(): ReactElement {
 						{ billing.isSuccess && formatCurrency( billing.data.costs.total, 'USD' ) }
 
-						{ ! billing.isSuccess && <TextPlaceholder /> }
+						{ ! billing.isSuccess && ( billing.isLoading ? <TextPlaceholder /> : '-' ) }
```
How about this (it should be easier to grok):
```jsx
{ billing.isSuccess && formatCurrency( billing.data.costs.total, 'USD' ) }

{ billing.isLoading && <TextPlaceholder /> }

{ billing.isError && <Gridicon icon="minus" /> }
```

atanas-dev reviewed (commented):
Works fine - I've just flagged a couple of improvements.

---

We don't need to wait for 3 retries in order to show this specific error. We can handle this with custom retry logic in useBillingDashboardQuery (via the retry react-query option) where it returns false if the error code matches (so it does not retry at all), else - returns false if failureCount >= 3

kallehauge committed:
Partner Portal: replace billing fallback dashes with  Gridicon

We do this to adhere to how the rest of the portal is showing "dashes".

kallehauge committed:
Partner Portal: replace billing ternary loading state with separate conditions

kallehauge committed:
Partner Portal: skip query retry on "no_billing_invoice_available" error

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/billing-summary/index.tsx:
```diff
@@ -84,7 +84,8 @@ export default function BillingSummary(): ReactElement {
 					{ billing.isSuccess && numberFormat( billing.data.licenses.total, 0 ) }
 
-					{ ! billing.isSuccess && <TextPlaceholder /> }
+					{ ! billing.isSuccess &&
+						( billing.isLoading ? <TextPlaceholder /> : <Gridicon icon="minus" /> ) }
```
Missed a spot :)

atanas-dev commented on client/jetpack-cloud/sections/partner-portal/billing-details/index.tsx:
```diff
@@ -119,12 +116,14 @@ export default function BillingDetails(): ReactElement {
 							} ) }
 
-						{ ! billing.isSuccess && <TextPlaceholder /> }
+						{ ! billing.isSuccess && billing.isLoading && <TextPlaceholder /> }
```
This should be reducible to just billing.isLoading, no?

atanas-dev reviewed (approved):
Added 2 super minor notes - feel free to merge after 👍

kallehauge committed:
Partner Portal: refactored isLoading status for billing dashboard

kallehauge committed:
Partner Portal: update billing query test to support retry logic

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6065496

Thank you @kallehauge for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6065496).
